Check Settings, Spec, and Coefficients Files

andkay opened the pull request:
This PR will address [#784](https://github.com/ActivitySim/activitysim/issues/784). 

The proposed approach includes some relatively simple "smoke test" of YAML, SPEC, and coefficients config files:
- The component settings are first loaded into the relevant Pydantic data model. 
- Then, the coefficients file is then read into memory, if defined at the top level of the data model.
- Then, the SPEC file is read into memory, similarly, if defined at the top level of the data model.
- Finally, one of the two available methods (`simulate.eval_coefficients` or `simulate.eval_nest_coefficients`) is run.

As scoped, validating expressions is not included, and the code isn't worried about tables at all.

Currently, there is a small subset of models that get tested, which include simpler configs, one with a preprocessor, and one with nested coefficients. Some basic logging is included. There may be more cases that need to be solved, but I did want to get some feedback / validation before pressing onward. I'll highlight some of my own thoughts below, but also want to get some eyes on this to make sure the contribution meets the issue-brief and will be useful.

**Autodiscovery of Model Settings YAML Files**
Currently, there is a dict that maps the names of component models to a Pydantic model class, as well as the relevant model file name. This will be pretty cumbersome to maintain.

A better way forward would be identify the default arguments to `model_settings` (Pydantic) and `model_settings_file_name` (YAML file) from the signature of each step. I'm not totally sure how to extract this information from the step wrapper around the callable.

**Error Handling**
An important design decision (which is still TBD) will be if any custom errors or warnings are raised by the `settings_checker` module (currently, this is not included. In my thinking, the main point of the settings checker is to surface any exceptions or warnings that would otherwise be raised at runtime earlier in the run process -- not necessarily to introduce any new error handling. If that is a decent assumption, responsibility for raising problems should be delegated upstream to the underlying Pydantic models and ActivitySim methods. 

**Nested Coefficients**
Currently, I am using `simulate.eval_nested_coefficients` to  evaluate models with the NESTS property. So far, this seems to behave as I expect -- but I am a little wary that it is misguided. With the settings checker turned off, I have not been been able to identify anything in the `prototype_mtc` model that actually calls this function.

**Calling the checker**
The checker is currently being called from `cli.run` for convenience of development. But I suspect that it needs to be migrated (or added) to the `State.run` method.

aletzdy committed:
fixing disaggregate accessibility bug in zone sampler

andkay committed:
feat: adds initial setting_checker module to abm/models

andkay committed:
chore: add .vscode settings to gitignore

andkay committed:
feat: adds first test case for prepopulating settings pydantic model (accessibility)

andkay committed:
feat: adds initial spec checker - wip

andkay committed:
feat: update settings checker to load spec and evaluate coefficients. add second model to validate

andkay committed:
fix: adds missing arg to eval_coefficients

andkay committed:
feat: adds atworksubtour_frequency to settings checker

andkay committed:
feat: moves loading spec and coefficients to independent functions

andkay committed:
refactor: renames main settings checker function. moves loading model settings to independent function

andkay committed:
refactor: moves load, spec, and coef eval checks to an independent function called from main checker. run formatter

andkay committed:
feat: adds load SPEC check for preprocessor settings

andkay committed:
chore: run formatter

andkay committed:
feat: initial checking for model setting with nested coefficients

jpn-- commented on activitysim/abm/models/settings_checker.py:
```diff
@@ -0,0 +1,158 @@
+
+def check_model_settings(state: State) -> None:
+
+    components = state.settings.models  # _RUNNABLE_STEPS.keys() may be better?
```
state.settings.models is probably the right thing here not `_RUNNABLE_STEPS`, as the latter includes every step that might be run, but the former includes only that which we want to run right now.  It's totally legit to have incomplete/invalid settings for steps we are not trying to run right now.

jpn-- commented on activitysim/abm/models/settings_checker.py:
```diff
@@ -0,0 +1,158 @@
+logger = logging.getLogger(__name__)
+file_logger = logger.getChild("logfile")
+
+COMPONENTS_TO_SETTINGS = {
```
This is probably a necessary evil for now.

andkay committed:
feat: adds additional model settings. force try_load_spec to return an empty dataframe, even if no top level spec exists

andkay committed:
feat: adds additional model settings to check. adds special handler for disaggregate accessibility (needs more testing)

i-am-sijia committed:
Revert "fixing disaggregate accessibility bug in zone sampler"

This reverts commit be5d093a972905d2fb4694fed6fa143cda58f581.

andkay committed:
feat: adds settings check for initalize_landuse

andkay committed:
refactor: reorders checking for joint tour composition settings

andkay committed:
feat: adds setting check for joint tour destination

andkay committed:
feat: adds settings check for joint tour frequency composition

andkay committed:
feat: adds settings check for joint tour frequency

andkay committed:
feat: adds settings check for joint tour participation

andkay committed:
feat: adds setting check for joint tour scheduling

andkay committed:
feat: adds settings checks for workplace location and school location

andkay committed:
feat: adds setting check for mandatory tour frequency

andkay committed:
feat: adds settings check for non mandatory tour destination

andkay committed:
feat: adds settings check for parking location choice

andkay committed:
feat: adds settings check for school escorting

andkay committed:
wip: adds stop_frequencies to components dict, but requires a separate flow for nested specs

andkay committed:
feat: adds setting checks for summarize

andkay committed:
feat: adds settings check for telecommute frequency

andkay committed:
feat: adds setting check for tour mode choice

andkay committed:
feat: adds setting check for tour od choice

andkay committed:
feat: adds setting check for tour scheduling probabilistic

andkay committed:
feat: adds setting check for transit pass ownership

andkay committed:
feat: adds setting check for transit pass subsidy

andkay committed:
feat: adds settings check for trip departure choice

andkay committed:
feat: adds setting check for trip destination

andkay committed:
feat: adds setting check for write trip matrices

andkay committed:
feat: adds setting check for trip mode choice

andkay committed:
feat: adds settings checks for trip purpose and destination

andkay committed:
feat: adds setting check for trip purpose

andkay committed:
feat: adds setting check for vehicle allocation

andkay committed:
feat: adds setting check for vehicle type choice

andkay committed:
feat: adds setting check for work from home

andkay commented:
Nearly all models in the `abm.models` directory now includes some level of config validation. More documentation to come shortly, but here is the short version:
- Some models do not have YAML settings files and are therefore excluded
- A few models are not configured in either of the SANDAG or MTC example models. They can't be tested to ensure the checks run sensibly and are therefore excluded. These include:
  - `tour_od_choice`
  - `tour_scheduling_probabilistic`
  - `trip_departure_choice` 
-  ~~The CDAP model is being skipped explicitly. I need to double check this one.~~
   - This was a spurious debug checkpoint and has been removed. 
- The `school_escorting` model  has only partial checks:
  - This model has a unique combination of different SPECs and Coefficent files for INBOUND, OUTBOUND, and OUTBOUND_COND that the general framework cannot handle.
- The `trip_scheduling_choice` model does have a corresponding YAML file in either the SANDAG or MTC models that I can test against, and so yields File Not Found errors.
- The `trip_scheduling` model is missing a coefficients file in both example models as well, which similarly yields File Not Found.
- Finally, the general settings file for the `stop_frequency` model IS included, but further checks are not implemented. This model has a unique NESTED_SPEC with different SPEC and Coefficients files defined in an array. This is not currently handled.

andkay committed:
docs: remove one comment

andkay committed:
refactor: removes spurious debug point around cdap spec load

andkay committed:
refactor: refactors setting checking to separate read/evaluate. checks now return result of tuple (Dataframe, Exception | None) for collection of errors.

andkay committed:
feat: adds setting check for intialize households

andkay committed:
feat: adds setting check for mandatory and non-mandatory tour scheduling

andkay committed:
feat: adds partial setting check for non-mandatory tour frequency (not fully implemented due to NESTED_SPEC)

andkay committed:
refactor: raise RunTimeError instead of custom

andkay committed:
feat: adds module-specific file handler to logging

andkay committed:
docs: remove one comment

andkay committed:
refactor: add more specific return type

andkay committed:
refactor: renames try_load_and_check to try_load_and_check_spec_coefs and directly injects model settings, separing check for reading  settings yaml files. useful for checking nested model settings

andkay committed:
feat: adds initial handling for SPEC_NEST configs. includes some prequisite refactors to error collection

andkay committed:
refactor: fix logging

andkay committed:
docs: update one comment

andkay committed:
refactor: remove commented raise statements used for debugging

andkay committed:
chore: run black formatter

andkay committed:
refactor: only ever evaluate spec/coef if both are available

jpn-- committed:
Restore input checker docs (#941)

* restore lost input checker docs
* render docs on workflow dispatch

dhensle committed:
Adding do_skim to tour mode choice (#943)

i-am-sijia committed:
SANDAG abm3 v15.0.0 contributions (#934)

* Reapply "fixing disaggregate accessibility bug in zone sampler"

This reverts commit de6d0ef2124e11654e2b7b3b66ba6f6520d2f9c3.

* blacken

* update the proto disagg accessibility regress file

due to fixing the zone sampler bug

andkay commented:
# Update for May 13th Demo

- Logging to stdout and file now included
- Errors are now collected into a single list, 
- SPEC Segments now being checked. Coefficient resolution for models with segmented specs are limited to those that have separate spec and coefficient files (not those models that use a general spec and PTYPE columns).

A few core limitations to the approached developed here:
1. Requirement to register a mapping of model names, Pydantic classes, and yaml files to the checker
1. Reliance on top-level SPEC and COEFFICIENT attributes. Things that sometimes appear but get skipped over include:
  a.  Alternative spellings: eg **SPECIFICATION**
  b. Case sensitive properties: e.g. **preprocessor** versus **PREPROCESSOR** (this is a little strange -- and I would recommend fixing this on the Pydantic side of the fence to create better consistency)
  c. Models with unique properties: e.g. **school_escorting**, which contains multiple pairs of SPEC/Coefficient files for Inbound/Outbound/Outbound Conditional

Other notes from previous comment [here](https://github.com/camsys/activitysim/pull/1#issuecomment-2868013111) continue to apply.

# To Do:
- Better error messages when SPEC/Coefficient labels can't be matched. Currently, the raised exception only reports out the offending index label, so this will need to be wrapped in a custom exception that includes the file name at a minimum.
- Per consortium feedback, there is a need to consider checks for processes that do get configured, but not part of the core abm.models module:
  - Some additional extension models (e.g. AV Ownership)
  - Network LOS settings
  - Constants
  - Shadow Pricing
- Possibly revisit Segmented Specs

# List of Models Being Checked

Below is the complete list of models being checked as of the most recent commit, relative to this comment.

Module | Model Checked
-- | --
accessibility.py | compute_accessibility
atwork_subtour_destination.py | atwork_subtour_destination
atwork_subtour_frequency.py | atwork_subtour_frequency
atwork_subtour_mode_choice.py | atwork_subtour_mode_choice
atwork_subtour_scheduling.py | atwork_subtour_scheduling
auto_ownership.py | auto_ownership_simulate
cdap.py | cdap_simulate
disaggregate_accessibility.py | compute_disaggregate_accessibility
free_parking.py | free_parking
initialize.py | initialize_landuse, initialize_households
joint_tour_composition.py | joint_tour_composition
joint_tour_destination.py | joint_tour_destination
joint_tour_frequency.py | joint_tour_frequency
joint_tour_frequency_composition.py | joint_tour_frequency_composition
joint_tour_participation.py | joint_tour_participation
joint_tour_scheduling.py | joint_tour_scheduling
location_choice.py | workplace_location, school_location
mandatory_scheduling.py | mandatory_tour_scheduling
mandatory_tour_frequency.py | mandatory_tour_frequency
non_mandatory_destination.py | non_mandatory_tour_destination
non_mandatory_scheduling.py | non_mandatory_tour_scheduling
non_mandatory_tour_frequency.py | non_mandatory_tour_frequency
parking_location_choice.py | parking_location
school_escorting.py | school_escorting
stop_frequency.py | stop_frequency
summarize.py | summarize
telecommute_frequency.py | telecommute_frequency
tour_mode_choice.py | tour_mode_choice_simulate
tour_od_choice.py | tour_od_choice
tour_scheduling_probabilistic.py | tour_scheduling_probabilistic
transit_pass_ownership.py | transit_pass_ownership
transit_pass_subsidy.py | transit_pass_subsidy
trip_departure_choice.py | trip_departure_choice
trip_destination.py | trip_destination
trip_matrices.py | write_trip_matrices
trip_mode_choice.py | trip_mode_choice
trip_purpose.py | trip_purpose
trip_purpose_and_destination.py | trip_purpose_and_destination
vehicle_allocation.py | vehicle_allocation
vehicle_type_choice.py | vehicle_type_choice
work_from_home.py | work_from_home

andkay commented:
Also todo (and probably important). The functions that try to load configurations are designed to consistently return a pd.DataFrame object, using an empty DataFrame when valid configuration data cannot be loaded. 

This design was necessary initially. Irrespective of success/failure, the results were being shipped to downstream checks for evaluating the coefficient and spec data - which requires the `df.copy()` interface (and handling the no method AttributeError at that point seemed kind of spurious).

I have since changed things to explicitly only try evaluation when both data sets are available. Changing the return type to `Optional[DataFrame]` and using `None` as the failure return type is probably the safer way forward, because there does exist the possibility that a CSV file is present in the configs directories but blank, which could slip through the cracks.

jpn-- committed:
Update to use pandas v2.* (#932)

* updates for pandas 2.2

* pytables 3.9

* input checker message failbacks

* fix veh type categoricals

* restore original pandas read_csv NaNs

* is_monotonic_increasing

* fix disagg acc sorting

* drop unused indexes

* update pipeline ref

* temporarily disable sharrow in vehicle alloc

* fix dtype problem

* ensure MAX index does not overflow

* sort on join to preserve index ordering from old pandas

* local compute test simplifies debugging

* more robust conversion to pyarrow

* rewrite df.eval to fast_eval

* change xarray pin

* fix zarr pin

* update numpy and dask pins

* wrap raw fast_eval in pd.Series

* don't skip sharrow in veh alloc

* rebuild ref pipeline

* make fast_eval more robust

* revise external targets

* prefer public API

* Update activitysim-dev-base.yml

restore accidentally removed larch

* add note about why fast_eval exists and how to undo it

jpn-- committed:
pin larch<6 for now (#947)

dkyleward committed:
Minor typo/bug fixes (#945)

* Fix typos in summarize docstring

* Fix log message in util.delete_files

* Fix docstring for VehicleAllocationSettings

---------

Co-authored-by: Jeffrey Newman <jeff@driftless.xyz>

andkay committed:
chore: better logging in model settings load

andkay committed:
feat: make setting checker optional

andkay committed:
feat: adds setting check for input checker

andkay committed:
feat: adds setting check for shadow pricing yaml

andkay committed:
chore: remove some outdated comments

andkay committed:
feat: adds setting checks for intialize los

andkay committed:
refactor: changes return type from empty df -> None when spec/coefs are unavaialable

andkay committed:
feat: adds logic for checking settings with templated coefficients

andkay committed:
chore: run formatter

andkay committed:
chore: fix alphabetization of components registry

andkay committed:
feat: check settings for write_data_dictionary

andkay committed:
feat: allow for detailed checks for SPEC_SEGMENTS with PTYPE against main model spec

andkay committed:
feat: allow for arbitrary loading of spec files from any subsettings in model settings

andkay committed:
feat: allow custom checks for unusual spec/coefficient pairs

andkay committed:
feat: adds custom SettingCheckerError exception for improved logging

andkay committed:
refactor: better checks and warnings when SPEC or COEFFICENT fields are
missing values.

because the fields in the settings classes are inherited, it is not
usually possible for the settings checker to determine if a path to an
external file is actually required for a particular model component.

the workaround is to issue a specific warning that a path *may* be
expected and users should check the YAML file.

the ultimate solution would be to define more robust Pydantic data
models that ensure that fields are marked as required when appropriate.

andkay committed:
chore: fix spelling.

andkay committed:
refactor: renames COMPONENTS_TO_SETTINGS -> CHECKER_SETTINGS and feeds to checker entrypoint as argument.

andkay committed:
refactor: single entry for settings checker. import extension check settings as dict.

andkay committed:
chore: run black formatter

andkay committed:
refactor: remove commented code

andkay committed:
fix: skip errors for write_data_dictionary if optional yaml file is not loadable

andkay committed:
chore: run black formatter

dhensle committed:
Preprocessing & Annotation functionality (#957)

* initial batch of models with pre and post processing added

* nonmand sched and transit pass models

* first pass at preprocessing and annotate functionality in all models

* fixing bugs in jtf and trip purpose

* adding persons back in to locals_d in jtc

* model name missing in tour scheduling

* missing expressions import in tour sched prob

* ci unit tests & fixing estimation test error

* addressing review comments

---------

Co-authored-by: Ali Etezady <58451076+aletzdy@users.noreply.github.com>

i-am-sijia committed:
Implement UV for dependency lock (#956)

* first uv sync

* explicitly set np.int64 dtype before calling pd set_index()

Pandas 2 got rid of the `Int64Index` class, it was downcasting to int32 for the input land use table

* Add all dependencies from Conda environments

Consider removing unused ones at a later time

* Remove orca

This dependency was replaced with “state” for workflow orchestration.

* Update release instructions for `uv`

* Remove `--no-default-groups` from `uv sync` instructions

* Trial updating Github Action to uv instead of conda

* Bug fixes

* Try downgrading sharrow

* change core test to windows runner

* use win for all core tests

* index type mismatch

* formatting

* unlock sharrow to 2.14

* correct a typo

* Update Model Setup page in user guide

* Update Ways to Run the Model page

* Remove duplicates

* Simplify pyproject.toml for github action and use only this group

* Update remainder of user guide docs

* Update dev guide [makedocs]

* Update lockfile from changes to dependencies

* Remove conda from other github actions

* Debug doc building [makedocs]

* Update install instructions [makedocs]

* update installation instructions [makedocs]

* notes on uv options [makedocs]

* address review comment

---------

Co-authored-by: Josie Kressner <josiekressner@gmail.com>

jpn-- committed:
Merge branch 'main' into settings-checker

andkay committed:
fix(tests): remove unused COEFFICIENTS from MWCOG test model

andkay committed:
fix(tests): re-add deleted test_mtc.py test script

andkay committed:
Merge branch 'settings-checker' of https://github.com/camsys/activitysim into settings-checker

andkay committed:
fix: make model registry for checking a copy

jpn-- committed:
Skim name conflicts (#939)

* identify and flag naming conflicts in 2d vs 3d skims

* add skims doc

* run on workflow_dispatch

* add link to docs

* write error message in exception also

* add error message example

* code block formatting

* unit tests

andkay committed:
fix(test): remove unused general SPEC in non_mandatory_tour_scheduling.yaml from MWCOG example model

i-am-sijia committed:
Pandas 2 `assign_in_place()` fix with categoricals (#948)

* update cat columns in df when new cat exists

* blacken

* sort two categoricals before union

dhensle committed:
PNUM hard-coded and Vehicle table creation bug fix (#953)

* minor fixes for SimOR development

* blacken

* Using PNUM if available in JTFC

Required to maintain backwards compatibility in tour indexes in SANDAG models

---------

Co-authored-by: Jeffrey Newman <jeff@driftless.xyz>

jpn-- committed:
Merge branch 'main' into settings-checker

dhensle committed:
Minor Fixes: optional logsums, duplicate data in parking lot choice, activitysim version to logfile (#963)

* adding activitysim version to log file

* fix duplicate columns in parking location choice as described in #633

* making tour mode choice logsum optional in destionation choice models #847

* logsum settings in location choice, trip dest, tour_od

* optional logsums in trip destination

* blacken

---------

Co-authored-by: Ali Etezady <58451076+aletzdy@users.noreply.github.com>

jpn-- committed:
Merge branch 'main' into settings-checker

andkay commented:
Closed by https://github.com/ActivitySim/activitysim/pull/950
